### PR TITLE
Feature/Enhancement: Optional acceleration camera offset

### PR DESCRIPTION
**Pull request by RisingLeaf**
## Feature Details
This PR implements the following feature: Everytime the players flagship is thrusting, an offset is applied to the camera, to make the acceleration more "feelable". This is optional and can be turned on or off in the settings.
Added: a reversed setting to have the offset in the other direction

## UI Screenshots
https://github.com/RisingLeaf/RisingLeaf/blob/main/Screen%20Recording%202023-02-19%20at%204.02.27%20PM.mov

## Testing Done
See the video

## Performance Impact
Very, very slightly increased CPU usage

### CODE REVIEW

**Comment by RisingLeaf** [comment]
> I would call this "Camera Acceleration," and reserve panning for the ability to actually move the camera with the mouse. Otherwise, looks great!

done

File: source/Engine.cpp
Comment (Hurleveur) [inline source/Engine.cpp:460]: This line is too long

File: source/Engine.cpp
Comment (Hurleveur) [inline source/Engine.cpp:1492]: 1.0 is the same as 1. (and is a double btw)

File: source/Engine.cpp
Comment (Hurleveur) [inline source/Engine.cpp:1496]: Too long as well I recon (I disagree with that rule but hey I dont make the rules here)

File: source/Engine.cpp
Comment (Hurleveur) [inline source/Engine.cpp:1494]: Compare offsetMultiplier with 0. instead of 0 to avoid a needless convertion from int to double

File: source/Engine.cpp
Comment (Hurleveur) [inline source/Engine.cpp:1495]: Same this 20 should probably be a 20. as I believe Velocity is a double too

File: source/Engine.h
Comment (Hurleveur) [inline source/Engine.h:192]: Here as well no need to write 0.0 as 0. is the same

File: source/Preferences.cpp
Comment (Hurleveur) [inline source/Preferences.cpp:60]: Generally what I do with new features is try to keep the default option as it was before.
Its by no means a hard rule but I'd like some extra opinions on this; should this be on by default? I cant view the video from phone so I cant say

**Review by Hurleveur — CHANGES_REQUESTED** [review]
Impressive how simple it is on the code side to do this; the 0.0025 should probably be a const static variable though, so its clear what it is and how to change it. We could even have this setting with multiple speeds of how its affected (0.001, 0.002....) with a few values but I dont mind it this way.
I suppose nothing weird happens in 2x mode?
For the code itself:
Today's lesson: lines of maximum length of 100 and 0. is a double x)

File: source/Engine.cpp
Comment (RisingLeaf) [reply source/Engine.cpp]: done

File: source/Engine.cpp
Comment (RisingLeaf) [reply source/Engine.cpp]: done

File: source/Engine.cpp
Comment (RisingLeaf) [reply source/Engine.cpp:1499]: done

File: source/Engine.cpp
Comment (RisingLeaf) [reply source/Engine.cpp]: done

File: source/Engine.cpp
Comment (RisingLeaf) [reply source/Engine.cpp]: done

File: source/Engine.h
Comment (RisingLeaf) [reply source/Engine.h]: done

File: source/Preferences.cpp
Comment (RisingLeaf) [reply source/Preferences.cpp]: done

File: source/Engine.cpp
Comment (RisingLeaf) [reply source/Engine.cpp]: done

**Comment by RisingLeaf** [comment]
> I suppose nothing weird happens in 2x mode? 

Nothing I noticed

**Comment by Zitchas** [comment]
I was going to suggest defaulting this to "off", but actually it would be nice to have a "first run accessibility" screen with options to set this and the parallax on and off. I know some people who find the camera acceleration effect causes verify and other negative effects, and others who can't handle the parallax.

**Comment by RisingLeaf** [comment]
> I was going to suggest defaulting this to "off", but actually it would be nice to have a "first run accessibility" screen with options to set this and the parallax on and off. I know some people who find the camera acceleration effect causes verify and other negative effects, and others who can't handle the parallax.

I am not sure if I can do this, sounds out of the scope of this PR

**Comment by RisingLeaf** [comment]
> Instead of just an on/off toggle in the options, what if you could toggle between Normal, Off, and Inverted (which was discussed on discord)?

I could do this, let me finsih my model then I will get to this

**Comment by RisingLeaf** [comment]
> > Instead of just an on/off toggle in the options, what if you could toggle between Normal, Off, and Inverted (which was discussed on discord)?
> 
> I could do this, let me finsih my model then I will get to this

implemented

File: source/Preferences.cpp
Comment (Hurleveur) [inline source/Preferences.cpp:258]: Isnt this too long?

File: source/Engine.h
Comment (Hurleveur) [inline source/Engine.h:193]: You so you know I meant this one as a part of the function, but this is good for me as well

**Review by Hurleveur — APPROVED** [review]
Nice!
Looks like you already know how to handle preferences with more than 2 options.

File: source/Preferences.cpp
Comment (RisingLeaf) [reply source/Preferences.cpp:258]: No, my editor shows me 98

Comment (Hurleveur) [reply source/Preferences.cpp:258]: Okay you can resolve both comments

**Comment by Ferociousfeind** [comment]
The linear interpolation between "full camera swaying" and "camera fully centers the player's flagship" was bothering me, so I made a small change (switching from `offset + offsetChange` and `offset - offsetChange` to `1. - (1. - offset) * offsetChange` and `offset * offsetChange` (with offsetChange changed from 0.0025 to 0.996) and it looks more like this...

when the camera setting is set to "on"
https://user-images.githubusercontent.com/35383563/195194399-f76c2e6f-4323-4b35-9d9a-1e7a7e7dfe50.mp4

when it's set to "reversed"
https://user-images.githubusercontent.com/35383563/195194458-1b8f7cc8-e137-4c60-8967-20c243129615.mp4

It still has unpleasant snapping to it, but I don't think I know the code enough to implement exactly what I want (the camera using this "multiply the offset by a number less than 1" interpolation method) all on my own

**Comment by RisingLeaf** [comment]
> The linear interpolation between "full camera swaying" and "camera fully centers the player's flagship" was bothering me, so I made a small change (switching from `offset + offsetChange` and `offset - offsetChange` to `1. - (1. - offset) * offsetChange` and `offset * offsetChange` (with offsetChange changed from 0.0025 to 0.996) and it looks more like this...
> 
> when the camera setting is set to "on" https://user-images.githubusercontent.com/35383563/195194399-f76c2e6f-4323-4b35-9d9a-1e7a7e7dfe50.mp4
> 
> when it's set to "reversed" https://user-images.githubusercontent.com/35383563/195194458-1b8f7cc8-e137-4c60-8967-20c243129615.mp4
> 
> It still has unpleasant snapping to it, but I don't think I know the code enough to implement exactly what I want (the camera using this "multiply the offset by a number less than 1" interpolation method) all on my own

As a reward for your work , i will try to fully implement this

File: source/Engine.cpp
Comment (quyykk) [inline source/Engine.cpp:1503]: Instead of storing offset, why not store the `newCenter` as a data object? So that you can assign the new center instead of doing the double ternary thingy twice.

File: source/Engine.h
Comment (quyykk) [inline source/Engine.h:193]: This should probably go in Engine.cpp in the anonymous namespace with the other constants.

File: source/Preferences.cpp
Comment (quyykk) [inline source/Preferences.cpp:263]:
```suggestion
const string &Preferences::CameraAcceleration()
```

**Review by quyykk — COMMENTED** [review]
The acceleration camera should probably be disabled while hyperdrive jumping somewhere, cause it that case my flagship just disappears off-screen.

File: source/Engine.h
Comment (RisingLeaf) [reply source/Engine.h:193]: will do this

File: source/Engine.cpp
Comment (RisingLeaf) [reply source/Engine.cpp:1503]: will do this

File: source/Engine.h
Comment (RisingLeaf) [reply source/Engine.h]: done

File: source/Engine.cpp
Comment (RisingLeaf) [reply source/Engine.cpp]: I am correcting myself, center and newCenter have to stay separated.
We would have done this earlier if it would work and it wont work.

Comment (quyykk) [reply source/Engine.cpp]: > I am correcting myself, center and newCenter have to stay separated.

Separated yes, but I mean storing `newCenter` as a data member so that it can be used to update `center` in Step.

Comment (RisingLeaf) [reply source/Engine.cpp]: ok

Comment (RisingLeaf) [reply source/Engine.cpp]: done

File: source/Engine.cpp
Comment (Amazinite) [inline source/Engine.cpp:1521]:
```suggestion
		bool thrust = flagship->IsThrusting();
		offsetMultiplier = offsetMultiplier >= thrust ? thrust
				: offsetMultiplier + OFFSET_CHANGE * (thrust ? 1 : -1);
```

**Comment by RisingLeaf** [comment]
Picking this back up cause I like it

**Comment by RisingLeaf** [comment]
resolved conflicts, tested and works!

**Review by Amazinite — COMMENTED** [review]
This new setting needs a tooltip.

**Comment by Koranir** [comment]
Personally, and no offense, but I don't like the way this is implemented right now. It is _far_ too jarring when the camera's acceleration starts and stops without regard to actual ship, and the values used feel quite off as well.

This kind of feature _is_ something I've looked into before with better results, see:

https://github.com/endless-sky/endless-sky/assets/101683475/29d13936-de7f-457b-a298-3a3c84ba7b0d

It's a lot smoother of an effect and (in my opinion) helps a lot more with the feeling of acceleration as opposed to yours:

https://github.com/endless-sky/endless-sky/assets/101683475/5983677e-601e-46af-b3ba-15503cabceb2

**Comment by Hurleveur** [comment]
Isnt that just a case of different variables for how quickly it moves? If so cant we upgrade this later with different options (smooth, quick etc)

**Comment by RisingLeaf** [comment]
Smoothed it out and adjusted some things.
https://github.com/endless-sky/endless-sky/assets/85687254/1dd48f9e-0ff7-45e0-b939-490505b60f81

I am pretty happy with this now, I like it.
But other opinions welcome.
I will say that this is a very performant and not too complex approach though.

**Comment by RisingLeaf** [comment]
> This new setting needs a tooltip.

Done as well as your other reviews

File: source/Preferences.cpp
Comment (Amazinite) [inline source/Preferences.cpp:59]:
```suggestion
```

File: source/Preferences.cpp
Comment (Amazinite) [inline source/Preferences.cpp:638]:
```suggestion


```

File: data/_ui/tooltips.txt
Comment (Amazinite) [inline data/_ui/tooltips.txt:1222]:
```suggestion
	`Toggle the dynamic camera that reacts to your ships acceleration between on, reversed, or off, where reversed is an offset in the other direction of the acceleration.`
```

File: source/Engine.cpp
Comment (Amazinite) [inline source/Engine.cpp:1590]:
```suggestion
		// Code block to calculate the center of the viewport.
		// If Camera Acceleration is enabled an offset calculated through the
		// flagship's velocity will be added to the center.
		newCenter = flagship->Center();
		double prefMul = Preferences::CameraAcceleration() == "on" ? 1.
			: (Preferences::CameraAcceleration() == "reversed" ? -1. : 0.);
		if(prefMul != 0. && !flagship->IsHyperspacing())
		{
			// If the flagship isn't thrusting, because no drag is experienced,
			// gently slide the flagship back to the center of the screen.
			if(flagship->IsThrusting() || flagship->IsReversing())
				offset += (flagship->Velocity() - offset) / 64.;
			else
				offset -= offset.Unit() / 32.;

			newCenter += offset * prefMul * 8;
		}
```

File: source/Engine.cpp
Comment (Amazinite) [inline source/Engine.cpp:1590]:
```suggestion
			newCenter += offset * prefMul * 8.;
```

File: source/Engine.cpp
Comment (Amazinite) [inline source/Engine.cpp:1580]:
```suggestion
		const string &pref = Preferences::CameraAcceleration();
		double prefMul = pref  == "on" ? 1. : (pref  == "reversed" ? -1. : 0.);
```
Thoughts on renaming `prefMul` to `accelDirection`/`offsetDirection`?

**Comment by Amazinite** [comment]
Honestly don't see this as being something I'd ever have on. I'd like to gauge the interest for a feature like this before merging.

**Review by alexrovw — CHANGES_REQUESTED** [review]
I like the feature, and I would use it.

Looks like there is a bug, though. It's creating a ghost of my outline.
Screenshot 2023-12-30 at 10 33 10 AM

**Comment by alexrovw** [comment]
This is fun as hell, tbh. But please do fix the flagship outline bug. :)

https://github.com/endless-sky/endless-sky/assets/98854112/8c6dcbb3-9eff-4182-a68e-c6ee90d9f302

**Comment by alexrovw** [comment]
Oh yeah. This is loads of fun. I'm definitely using this. Please do fix the bug with the flagship outline, though. :)

https://github.com/endless-sky/endless-sky/assets/98854112/eeea60da-f5c2-44f0-aa4a-8ba61166ce12

**Comment by Hurleveur** [comment]
I may use it as well, could help in making the game feel more reactive
That said it should prob be off by default

File: source/Engine.cpp
Comment (Amazinite) [inline source/Engine.cpp:1067]: Well this comment isn't quite true anymore.

**Review by alexrovw — CHANGES_REQUESTED** [review]
Interesting... Now the outline seems to appear behind me at high levels of acceleration (Bellows on a Merganser).
Screenshot 2023-12-30 at 11 31 51 AM

**Review by alexrovw — CHANGES_REQUESTED** [review]
I tried the reversed mode for this setting and... oh god what is this?! It killed my game :(

https://github.com/endless-sky/endless-sky/assets/98854112/c46bd66e-d14c-4daa-89b2-8b5657dcd201

**Comment by RisingLeaf** [comment]
As long as this is up in the air, I hope it is understandable that I wont invest more time

File: source/Engine.cpp
Comment (RisingLeaf) [reply source/Engine.cpp:1068]:
```suggestion
```

**Comment by alexrovw** [comment]
Looks good! The previously described issues are both tested and verified fixed:

1. The outline matches the position of the ship in all of my test cases, including complex turning/stopping/reversing maneuvers
2. The ship no longer slingshots and freeze-locks the game when the reverse mode is enabled for this setting.

That said, I ran into one small edge case. I leave it up to you to decide if it's worth fixing or not. The acceleration effects do not apply correctly when the user is being accelerated because of knockback from their own weapon. This is most easily demonstrated on an Archon, but I also showed it on a Merganser with a Thorax Cannon. These are only illustrative of the issue and not actual cases the player will run into.

https://github.com/endless-sky/endless-sky/assets/98854112/ac04a5eb-4012-43cb-9d77-d471c5f436b9

https://github.com/endless-sky/endless-sky/assets/98854112/eb28bc35-ad12-4bcd-8eab-0e61a6c44a8d

Bearing in mind that these are edge cases, I'm happy to approve regardless. A future PR could potentially look at the relationship between knockback and acceleration.

**Review by alexrovw — APPROVED** [review]

**Comment by RisingLeaf** [comment]
I will say that knockback should come in another PR

**Comment by RisingLeaf** [comment]
Closing in favour of the other one.

**Pull request closed without merging**
